File: cpr-fantasy-football/lib/match-processor.ts
import { fetchCSV } from './data-fetcher';
import { CSV_URLS, MATCH_COLUMNS } from './constants';

export interface MatchPlayerPerformance {
  name: string;
  appearance: number;
  goals: number;
  assists: number;
  cleanSheet: number;
  yellowCard: number;
  redCard: number;
  points: number;
  mom1?: number;
  mom2?: number;
  mom3?: number;
  dod?: number;
}

export interface Match {
  date: string;
  team: 'CPR' | 'CPRA';
  opponent: string;
  score: string;
  cprScore: number;
  opponentScore: number;
  gameweek: string;
  players: MatchPlayerPerformance[];
  isSubmitted?: boolean; // Flag to indicate if this match data comes from captain submission
  matchSummary?: string;
  matchReport?: string;
}

/**
 * Parse score from match description (e.g., "CPR 3v2 Opponent" or "CPRA 4v4 Opponent")
 */
function parseScore(matchDescription: string): { team: 'CPR' | 'CPRA'; cpr: number; opponentScore: number; opponentName: string } {
  // Try to match CPRA first (more specific), then CPR
  const cpraMatch = matchDescription.match(/CPRA\s+(\d+)v(\d+)\s+(.+)/i);
  if (cpraMatch) {
    return {
      team: 'CPRA',
      cpr: parseInt(cpraMatch[1], 10),
      opponentScore: parseInt(cpraMatch[2], 10),
      opponentName: cpraMatch[3].trim(),
    };
  }

  const cprMatch = matchDescription.match(/CPR\s+(\d+)v(\d+)\s+(.+)/i);
  if (cprMatch) {
    return {
      team: 'CPR',
      cpr: parseInt(cprMatch[1], 10),
      opponentScore: parseInt(cprMatch[2], 10),
      opponentName: cprMatch[3].trim(),
    };
  }

  // Default to CPR if can't parse
  return {
    team: 'CPR',
    cpr: 0,
    opponentScore: 0,
    opponentName: matchDescription.replace(/^CPR(A)?\s+/i, '').trim()
  };
}

/**
 * Fetches and processes match data with player performances
 * @returns Array of matches sorted by date (most recent first)
 */
export async function getMatches(): Promise<Match[]> {
  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const matchData = await fetchCSV<any>(CSV_URLS.MATCH_DETAILS);

    // Skip first 3 rows (headers)
    const dataRows = matchData.slice(3);

    // Group by match (date + game description)
    const matchesMap = new Map<string, Match>();

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    dataRows.forEach((row: any) => {
      const date = String(row[MATCH_COLUMNS.DATE] || '').trim();
      const gameweek = String(row[MATCH_COLUMNS.GAMEWEEK] || '').trim();
      const matchDescription = String(row[MATCH_COLUMNS.GAME] || '').trim();
      const playerName = String(row[MATCH_COLUMNS.PLAYER] || '').trim();

      if (!date || !matchDescription || !playerName) return;

      const matchKey = `${date}-${matchDescription}`;

      // Create match entry if it doesn't exist
      if (!matchesMap.has(matchKey)) {
        const { team, cpr, opponentScore: oppScore, opponentName: oppName } = parseScore(matchDescription);
        matchesMap.set(matchKey, {
          date,
          team,
          opponent: oppName,
          score: `${cpr}-${oppScore}`,
          cprScore: cpr,
          opponentScore: oppScore,
          gameweek,
          players: [],
        });
      }

      const match = matchesMap.get(matchKey)!;

      // Add player performance using column constants
      match.players.push({
        name: playerName,
        appearance: Number(row[MATCH_COLUMNS.APPEARANCE]) || 0,
        goals: Number(row[MATCH_COLUMNS.GOALS]) || 0,
        assists: Number(row[MATCH_COLUMNS.ASSISTS]) || 0,
        cleanSheet: Number(row[MATCH_COLUMNS.CLEAN_SHEET]) || 0,
        yellowCard: Number(row[MATCH_COLUMNS.YELLOW_CARD]) || 0,
        redCard: Number(row[MATCH_COLUMNS.RED_CARD]) || 0,
        points: Number(row[MATCH_COLUMNS.TOTAL_POINTS]) || 0,
      });
    });

    // Convert to array and sort by date (most recent first)
    const matches = Array.from(matchesMap.values());
    matches.sort((a, b) => {
      // Parse dates (format: DD/MM/YYYY)
      const [dayA, monthA, yearA] = a.date.split('/').map(Number);
      const [dayB, monthB, yearB] = b.date.split('/').map(Number);
      const dateA = new Date(yearA, monthA - 1, dayA);
      const dateB = new Date(yearB, monthB - 1, dayB);
      return dateB.getTime() - dateA.getTime();
    });

    return matches;
  } catch (error) {
    console.error('Error processing matches:', error);
    throw error;
  }
}

File: conda-fantasy-football/src/match_processor.py
"""Match data processing."""

import pandas as pd
import re
from typing import Dict, List, Tuple
from .data_fetcher import fetch_csv, parse_number
from .config import CSV_URLS, MATCH_COLUMNS


def parse_score(match_description: str) -> Tuple[str, int, int, str]:
    """
    Parse score from match description (e.g., "CPR 3v2 Opponent" or "CPRA 4v4 Opponent").

    Args:
        match_description: Match description string

    Returns:
        Tuple of (team, cpr_score, opponent_score, opponent_name)
    """
    # Try to match CPRA first (more specific), then CPR
    cpra_match = re.search(r"CPRA\s+(\d+)v(\d+)\s+(.+)", match_description, re.IGNORECASE)
    if cpra_match:
        return (
            "CPRA",
            int(cpra_match.group(1)),
            int(cpra_match.group(2)),
            cpra_match.group(3).strip(),
        )

    cpr_match = re.search(r"CPR\s+(\d+)v(\d+)\s+(.+)", match_description, re.IGNORECASE)
    if cpr_match:
        return (
            "CPR",
            int(cpr_match.group(1)),
            int(cpr_match.group(2)),
            cpr_match.group(3).strip(),
        )

    # Default to CPR if can't parse
    return (
        "CPR",
        0,
        0,
        re.sub(r"^CPR(A)?\s+", "", match_description, flags=re.IGNORECASE).strip(),
    )


def get_matches() -> List[Dict]:
    """
    Fetch and process match data with player performances.

    Returns:
        List of match dictionaries sorted by date (most recent first)
    """
    # Match data uses generic column names like '_1', '_2' to match TypeScript behavior
    match_data = fetch_csv(CSV_URLS["MATCH_DETAILS"], use_generic_headers=True)

    # Skip first 3 rows (headers)
    data_rows = match_data.iloc[3:]

    # Group by match (date + game description)
    matches_map: Dict[str, Dict] = {}

    for _, row in data_rows.iterrows():
        date = str(row.get(MATCH_COLUMNS["DATE"], "")).strip()
        gameweek = str(row.get(MATCH_COLUMNS["GAMEWEEK"], "")).strip()
        match_description = str(row.get(MATCH_COLUMNS["GAME"], "")).strip()
        player_name = str(row.get(MATCH_COLUMNS["PLAYER"], "")).strip()

        if not date or not match_description or not player_name:
            continue

        match_key = f"{date}-{match_description}"

        # Create match entry if it doesn't exist
        if match_key not in matches_map:
            team, cpr_score, opp_score, opp_name = parse_score(match_description)
            matches_map[match_key] = {
                "date": date,
                "team": team,
                "opponent": opp_name,
                "score": f"{cpr_score}-{opp_score}",
                "cpr_score": cpr_score,
                "opponent_score": opp_score,
                "gameweek": gameweek,
                "players": [],
            }

        match = matches_map[match_key]

        # Add player performance
        match["players"].append(
            {
                "name": player_name,
                "appearance": parse_number(row.get(MATCH_COLUMNS["APPEARANCE"], 0)),
                "goals": parse_number(row.get(MATCH_COLUMNS["GOALS"], 0)),
                "assists": parse_number(row.get(MATCH_COLUMNS["ASSISTS"], 0)),
                "clean_sheet": parse_number(row.get(MATCH_COLUMNS["CLEAN_SHEET"], 0)),
                "yellow_card": parse_number(row.get(MATCH_COLUMNS["YELLOW_CARD"], 0)),
                "red_card": parse_number(row.get(MATCH_COLUMNS["RED_CARD"], 0)),
                "points": parse_number(row.get(MATCH_COLUMNS["TOTAL_POINTS"], 0)),
            }
        )

    # Convert to list
    matches = list(matches_map.values())

    # Sort by date (most recent first)
    def parse_date(date_str: str):
        """Parse DD/MM/YYYY format"""
        try:
            parts = date_str.split("/")
            if len(parts) == 3:
                day, month, year = map(int, parts)
                return (year, month, day)
        except:
            pass
        return (0, 0, 0)

    matches.sort(key=lambda m: parse_date(m["date"]), reverse=True)

    return matches


def get_match_result_badge(cpr_score: int, opponent_score: int) -> str:
    """
    Get result badge emoji for a match.

    Args:
        cpr_score: CPR score
        opponent_score: Opponent score

    Returns:
        Emoji string representing win/draw/loss
    """
    if cpr_score > opponent_score:
        return "✅"  # Win
    elif cpr_score < opponent_score:
        return "❌"  # Loss
    else:
        return "⚖️"  # Draw
